File: data_structures/avl_tree/matlab/test_avl_tree.m
% test_avl_tree.m

t = avl_tree('new');
assert(avl_tree(t, 'is_empty'));

for v = [5, 3, 7, 1, 4]
    t = avl_tree(t, 'insert', v);
end
assert(avl_tree(t, 'search', 3));
assert(~avl_tree(t, 'search', 6));
assert(isequal(avl_tree(t, 'inorder'), [1, 3, 4, 5, 7]));

t2 = avl_tree('new');
for v = [5, 3, 7]
    t2 = avl_tree(t2, 'insert', v);
end
assert(avl_tree(t2, 'min') == 3);
assert(avl_tree(t2, 'max') == 7);

t3 = avl_tree('new');
for i = 1:100
    t3 = avl_tree(t3, 'insert', i);
end
h = avl_tree(t3, 'height');
assert(h <= ceil(1.44 * log2(101)) + 1);
io = avl_tree(t3, 'inorder');
assert(isequal(io, 1:100));

[t2, ok] = avl_tree(t2, 'remove', 3);
assert(ok);
assert(isequal(avl_tree(t2, 'inorder'), [5, 7]));
[t2, ok] = avl_tree(t2, 'remove', 99);
assert(~ok);

t_empty = avl_tree('new');
try
    avl_tree(t_empty, 'min');
    error('expected error not thrown');
catch e
    assert(strcmp(e.identifier, 'avl_tree:underflow'));
end

disp('All tests passed.');

File: data_structures/avl_tree/python/test_avl_tree.py
import math
import pytest
from avl_tree import AvlTree


def test_new_is_empty():
    t = AvlTree()
    assert t.is_empty()
    assert t.size() == 0


def test_insert_and_search():
    t = AvlTree()
    for v in [5, 3, 7, 1, 4]:
        t.insert(v)
    assert t.search(3)
    assert not t.search(6)
    assert t.inorder() == [1, 3, 4, 5, 7]


def test_min_max():
    t = AvlTree()
    for v in [5, 3, 7]:
        t.insert(v)
    assert t.min() == 3
    assert t.max() == 7


def test_height_log_n():
    t = AvlTree()
    for i in range(1, 101):
        t.insert(i)
    h = t.height()
    expected = math.ceil(1.44 * math.log2(101))
    assert h <= expected + 1


def test_inorder_sorted():
    t = AvlTree()
    for i in range(1, 101):
        t.insert(i)
    io = t.inorder()
    assert io == list(range(1, 101))


def test_remove():
    t = AvlTree()
    for v in [5, 3, 7, 1, 4]:
        t.insert(v)
    assert t.remove(1)
    assert t.remove(3)
    assert t.inorder() == [4, 5, 7]
    assert not t.remove(99)


def test_min_empty_raises():
    with pytest.raises(IndexError):
        AvlTree().min()
